### PR TITLE
Ai summary table

### PR DESCRIPTION
**Pull request by shinuito**
Fixes:
When a game ends with consecutive passes, there's a console error which results from empty branches in the AI review having their (non existent) move coordinates checked. So added in a mv === undefined check.

## Proposed Changes
Adding in a new AISummaryTable component to the AIReview which tries to summarise the data according to score loss, and categorise the types of moves as Excellent = blue move, Great = one of the green moves, Good = some positive score change move or mistake <1 point, Inaccuracy <2 point mistake, Mistake <5 points loss, Blunder >=5 point loss.

  - It currently only shows for katago reviews (full or fast, but fast omits the Excellent and Great categories).
  - It currently only shows for moderators, in order to test and tune before showing it to everybody.
  - It should work on finished games and uploaded games, and now forked games.
  - Added a preference to remember if the user wants the table hidden or not and a toggle for that.

There's still an issue where if the user opens a game page and then navigates away quickly, there's a setState error ("Can't perform a React state update on an unmounted component."). A similar issues happens with PaginatedTable on profile pages (for moderators) and for the Tournaments list pages, but I don't know how to fix these issues yet.

### CODE REVIEW

**Comment by mmartin** [comment]
Regarding empty AI branches,  #1316 can be closed when this PR gets merged.

**Comment by BHydden** [comment]
> can be closed when this PR gets merged.

@mmartin FTR, GitHub will do this automatically if you type Fixes #1316 

- (`resolves` also works in place of `fixes`, and I think there might be some other key words that will trigger an auto close... `implents` maybe?)

File: src/views/Game/AIReview.tsx [hunk 1]
```diff
@@ -687,6 +709,220 @@ export class AIReview extends React.Component<AIReviewProperties, AIReviewState>
...
+            move_list.push(cur_move.player);
+            cur_move = cur_move.trunk_next;
+        }
+        return move_list;
+
+    }
+
+    private AiSummaryTableRowList(){
+        const summary_moves_list = [ ["", "", "", ""], ["", "", "", ""], ["", "", "", ""], ["", "", "", ""], ["", "", "", ""], ["", "", "", ""] ];
+        const ai_table_rows = [["Excellent"], ["Great"], ["Good"], ["Inaccuracy"], ["Mistake"], ["Blunder"]];
```
Comment (benjaminpjones) [inline src/views/Game/AIReview.tsx:727]: Would be good to internationalize these strings:

```
[[_("Excellent")], [_("Great")], [_("Good")], [_("Inaccuracy")], [_("Mistake")], [_("Blunder")]]
```

File: src/views/Game/AIReview.tsx [hunk 2]
```diff
@@ -959,6 +1194,11 @@ export class AIReview extends React.Component<AIReviewProperties, AIReviewState>
                         </span>
                     </div>
                 }
+                { (data.get("user").is_moderator && ((this.ai_review?.type === "full" ) || (this.ai_review?.type === "fast" )) && this.ai_review?.engine === "katago") &&
```
Comment (benjaminpjones) [inline src/views/Game/AIReview.tsx:1197]: Why check both 'full' and 'fast' here? I think those are the only two values `JGOFAIIReview.type` can have

File: src/views/Game/AIReview.tsx [hunk 3]
```diff
@@ -959,6 +1194,11 @@ export class AIReview extends React.Component<AIReviewProperties, AIReviewState>
...
+                <div>
+                    <AiSummaryTable headinglist = {["Type", "Black", "%", "White", "%"]} bodylist = {this.table_rows} avg_loss = {this.avg_score_loss} />
```
Comment (benjaminpjones) [inline src/views/Game/AIReview.tsx:1199]: Would be good to internationalize these headings too

File: src/views/Game/AIReview.tsx (hunk 1 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:727]: Sure, I assume that so they can be translated? Makes senes, I didn't know the syntax :)

Comment (BHydden) [reply src/views/Game/AIReview.tsx:727]: The two main ones are _("String") and [pgettext](https://github.com/online-go/online-go.com/blob/13fcc8b1e698eebb80b90790d6e8fe522c35dc8f/src/lib/translate.ts#L173) which allows you to add additional context for more complex strings

Comment (shinuito) [reply src/views/Game/AIReview.tsx:727]: Is there a simple way to just click to accept the suggested change?

File: src/views/Game/AIReview.tsx (hunk 2 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:1197]: So it's essentially just throwing in a "true" in there I suppose. I think initially I wasn't going to show the table for fast reviews, which was also a comment I had received. 

That is just because the data won't be as accurate, but it's the same with the AI graph itself and we do show that also. 

I guess we can take out that part.

File: src/views/Game/AIReview.tsx (hunk 3 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:1199]: I agree :)

File: src/views/Game/AIReview.tsx (hunk 1 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:727]:
```suggestion
        const ai_table_rows = [[_("Excellent")], [_("Great")], [_("Good")], [_("Inaccuracy")], [_("Mistake")], [_("Blunder")]];
```

File: src/views/Game/AIReview.tsx (hunk 2 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:1197]:
```suggestion
                { (data.get("user").is_moderator && this.ai_review?.engine === "katago") &&
```

File: src/views/Game/AIReview.tsx (hunk 3 above)
Comment (shinuito) [reply src/views/Game/AIReview.tsx:1199]:
```suggestion
                    <AiSummaryTable headinglist = {[_("Type"), _("Black"), "%", _("White"), "%"]} bodylist = {this.table_rows} avg_loss = {this.avg_score_loss} />
```

File: src/views/Game/AIReview.tsx (hunk 1 above)
Comment (benjaminpjones) [reply src/views/Game/AIReview.tsx]: Thanks for making that change!  Yeah basically there's a script that picks up all the _(...) and pgettext(...) and sends them to translate.online-go.com for translation.  Once the translators do their thing, the translations get added to the dictionary and these functions actually convert the text to the target language.

Pro tip: if you change your language to *Debug* in the settings, it shows you which strings have already been internationalized.  Strings wrapped in [...] are ready:

Screen Shot 2021-12-16 at 9 24 19 AM

Comment (shinuito) [reply src/views/Game/AIReview.tsx]: That sounds like a really convenient idea, to have a debug language like that :)

**Comment by anoek** [comment]
This is great, thanks for making this happen @shinuito :)

For that toggle, how do y'all feel about moving it up onto the key moves line?

image

Maybe instead of the "Hide summary table" text we could just have a little table icon precede it, such as `<i className="fa fa-table" />` or something?

The other thing that might be nice is if the table populated while the results were being streamed in, is that something that could be done relatively easily, or would that be a pain?

**Comment by shinuito** [comment]
> For that toggle, how do y'all feel about moving it up onto the key moves line?

That probably makes sense for saving space yes. Let me catch the branch up to devel and try that.

> Maybe instead of the "Hide summary table" text we could just have a little table icon precede it, such as `<i className="fa fa-table" />` or something?

I didn't actually think of using the fa icons, but that also seems like a good idea.

> The other thing that might be nice is if the table populated while the results were being streamed in, is that something that could be done relatively easily, or would that be a pain?

I was relying on the scores being the difference two consecutive entries so I'll have a think about it in an hour or so when I can look at it properly. Maybe something like looping through the existing entries by their keys for the full review, but I'll have to see how the scores array gets populated for the fast reviews.

It does seem like it would be nice to have something going on while the AI review is loading :)

**Comment by anoek** [comment]
> I was relying on the scores being the difference two consecutive entries
That makes sense, I don't think you need to rethink that, but maybe you could do something like count them if you have both entries, and if you're missing the prior entry just count it in a "pending" number or something? It'd make the display more lively and folks definitely like that :)

**Comment by shinuito** [comment]
@anoek I've updated that now. I've moved the toggle beside the "Key Moves" and used fa-table. I think it looks better.



I also have tried to make the table display in real time the data being populated, and it has an extra row called "Moves pending" that counts how many data points are missing and then disappears when it hits zero.

I suppose that could be styled into anything really, a loading bar etc.

Let me know what you think, or if you have any more suggestions :)

**Comment by anoek** [comment]
Looking great! A loading bar would be awesome, not a requirement by any stretch but if you feel like it, that would be slick :)

**Comment by shinuito** [comment]
So what do you think of something like this then?



That's just using the basic `<progress>` html element and passing the right values to it. I guess one thing is that left to itself it might be styled by the browser. I suppose one could override that with some custom styles but I'm not sure what would be the best.

I was hoping to let some of the mods try some things out with the table, suggest fixes and tweaks (like if they want to change any of the thresholds for good inaccuracy etc), then remove the `is_moderator` to let everyone see it once we're happy with it. I think there was some requests for other metrics that's might be interesting, but that's an ongoing discussion and maybe some future updates could be kept for mod usage so as not to overload the users with information.

**Comment by anoek** [comment]
Awesome work @shinuito , i'll get it up in a sec
